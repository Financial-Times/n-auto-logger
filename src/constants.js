export const CATEGORIES = {
	FETCH_RESPONSE_OK: 'FETCH_RESPONSE_OK',
	FETCH_RESPONSE_ERROR: 'FETCH_RESPONSE_ERROR',
	FETCH_NETWORK_ERROR: 'FETCH_NETWORK_ERROR',
	NODE_SYSTEM_ERROR: 'NODE_SYSTEM_ERROR',
	CUSTOM_ERROR: 'CUSTOM_ERROR',
};

export const RESULTS = {
	SUCCESS: 'success',
	FAILURE: 'failure',
};

export default {
	CATEGORIES,
	RESULTS,
};
