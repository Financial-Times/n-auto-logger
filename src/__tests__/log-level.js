import logger, { logAction, logOperation } from '../index';
import { LOG_LEVELS } from '../constants';

jest.mock('@financial-times/n-logger');

const commonErrorInstance = { status: 404, message: 'Not Found' };

const runCommonTestCases = () => {
	describe('success of', () => {
		it('async function with async sub actions', async () => {
			const callFunction = () => Promise.resolve('foo');
			const operationFunction = async meta => {
				await logAction(callFunction)(null, meta);
			};
			const enhanced = logOperation(operationFunction);
			await enhanced();
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('non-async function with non async sub actions', async () => {
			const callFunction = () => {};
			const operationFunction = meta => {
				logAction(callFunction)(null, meta);
			};
			const enhanced = logOperation(operationFunction);
			await enhanced();
			expect(logger.info.mock.calls).toMatchSnapshot();
		});
	});

	describe('failure of', () => {
		it('non-async function with non-async sub actions', async () => {
			const callFunction = () => {
				throw commonErrorInstance;
			};
			const operationFunction = meta => {
				try {
					logAction(callFunction)(null, meta);
				} catch (e) {
					throw e;
				}
			};
			const enhanced = logOperation(operationFunction);
			try {
				await enhanced();
			} catch (e) {
				expect(logger.info.mock.calls).toMatchSnapshot();
				expect(logger.warn.mock.calls).toMatchSnapshot();
				expect(logger.error.mock.calls).toMatchSnapshot();
			}
		});

		it('async function with async sub actions', async () => {
			const callFunction = async () => {
				throw commonErrorInstance;
			};
			const operationFunction = async meta => {
				try {
					await logAction(callFunction)(null, meta);
				} catch (e) {
					throw e;
				}
			};
			const enhanced = logOperation(operationFunction);
			try {
				await enhanced();
			} catch (e) {
				expect(logger.info.mock.calls).toMatchSnapshot();
				expect(logger.warn.mock.calls).toMatchSnapshot();
				expect(logger.error.mock.calls).toMatchSnapshot();
			}
		});
	});
};

describe('AUTO_LOG_LEVEL', () => {
	afterEach(() => {
		jest.resetAllMocks();
	});

	describe('=verbose logs every stage of operation and action', () => {
		beforeAll(() => {
			process.env.AUTO_LOG_LEVEL = LOG_LEVELS.verbose;
		});

		afterAll(() => {
			delete process.env.AUTO_LOG_LEVEL;
		});

		runCommonTestCases();
	});

	describe('=concise logs success/failure of operation and only failure of action ', () => {
		beforeAll(() => {
			process.env.AUTO_LOG_LEVEL = LOG_LEVELS.concise;
		});

		afterAll(() => {
			delete process.env.AUTO_LOG_LEVEL;
		});

		runCommonTestCases();
	});

	describe('=error logs only failure of operation and action', () => {
		beforeAll(() => {
			process.env.AUTO_LOG_LEVEL = LOG_LEVELS.error;
		});

		afterAll(() => {
			delete process.env.AUTO_LOG_LEVEL;
		});

		runCommonTestCases();
	});
});
