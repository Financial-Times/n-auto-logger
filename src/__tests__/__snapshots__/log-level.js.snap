// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AUTO_LOG_LEVEL =concise logs success/failure of action without param  failure of async function with async sub actions 1`] = `Array []`;

exports[`AUTO_LOG_LEVEL =concise logs success/failure of action without param  failure of async function with async sub actions 2`] = `
Array [
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =concise logs success/failure of action without param  failure of async function with async sub actions 3`] = `Array []`;

exports[`AUTO_LOG_LEVEL =concise logs success/failure of action without param  failure of non-async function with non-async sub actions 1`] = `Array []`;

exports[`AUTO_LOG_LEVEL =concise logs success/failure of action without param  failure of non-async function with non-async sub actions 2`] = `
Array [
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =concise logs success/failure of action without param  failure of non-async function with non-async sub actions 3`] = `Array []`;

exports[`AUTO_LOG_LEVEL =concise logs success/failure of action without param  success of async function with async sub actions 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =concise logs success/failure of action without param  success of non-async function with non async sub actions 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =error logs only failure of action failure of async function with async sub actions 1`] = `Array []`;

exports[`AUTO_LOG_LEVEL =error logs only failure of action failure of async function with async sub actions 2`] = `
Array [
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =error logs only failure of action failure of async function with async sub actions 3`] = `Array []`;

exports[`AUTO_LOG_LEVEL =error logs only failure of action failure of non-async function with non-async sub actions 1`] = `Array []`;

exports[`AUTO_LOG_LEVEL =error logs only failure of action failure of non-async function with non-async sub actions 2`] = `
Array [
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =error logs only failure of action failure of non-async function with non-async sub actions 3`] = `Array []`;

exports[`AUTO_LOG_LEVEL =error logs only failure of action success of async function with async sub actions 1`] = `Array []`;

exports[`AUTO_LOG_LEVEL =error logs only failure of action success of non-async function with non async sub actions 1`] = `Array []`;

exports[`AUTO_LOG_LEVEL =standard logs success/failure of operation and action  failure of async function with async sub actions 1`] = `Array []`;

exports[`AUTO_LOG_LEVEL =standard logs success/failure of operation and action  failure of async function with async sub actions 2`] = `
Array [
  Array [
    Object {
      "action": "callFunction",
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =standard logs success/failure of operation and action  failure of async function with async sub actions 3`] = `Array []`;

exports[`AUTO_LOG_LEVEL =standard logs success/failure of operation and action  failure of non-async function with non-async sub actions 1`] = `Array []`;

exports[`AUTO_LOG_LEVEL =standard logs success/failure of operation and action  failure of non-async function with non-async sub actions 2`] = `
Array [
  Array [
    Object {
      "action": "callFunction",
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =standard logs success/failure of operation and action  failure of non-async function with non-async sub actions 3`] = `Array []`;

exports[`AUTO_LOG_LEVEL =standard logs success/failure of operation and action  success of async function with async sub actions 1`] = `
Array [
  Array [
    Object {
      "action": "callFunction",
      "operation": "operationFunction",
      "result": "success",
    },
  ],
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =standard logs success/failure of operation and action  success of non-async function with non async sub actions 1`] = `
Array [
  Array [
    Object {
      "action": "callFunction",
      "operation": "operationFunction",
      "result": "success",
    },
  ],
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =verbose logs every stage of operation and action failure of async function with async sub actions 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
  Array [
    Object {
      "action": "callFunction",
      "operation": "operationFunction",
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =verbose logs every stage of operation and action failure of async function with async sub actions 2`] = `
Array [
  Array [
    Object {
      "action": "callFunction",
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =verbose logs every stage of operation and action failure of async function with async sub actions 3`] = `Array []`;

exports[`AUTO_LOG_LEVEL =verbose logs every stage of operation and action failure of non-async function with non-async sub actions 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
  Array [
    Object {
      "action": "callFunction",
      "operation": "operationFunction",
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =verbose logs every stage of operation and action failure of non-async function with non-async sub actions 2`] = `
Array [
  Array [
    Object {
      "action": "callFunction",
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =verbose logs every stage of operation and action failure of non-async function with non-async sub actions 3`] = `Array []`;

exports[`AUTO_LOG_LEVEL =verbose logs every stage of operation and action success of async function with async sub actions 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
  Array [
    Object {
      "action": "callFunction",
      "operation": "operationFunction",
    },
  ],
  Array [
    Object {
      "action": "callFunction",
      "operation": "operationFunction",
      "result": "success",
    },
  ],
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
    },
  ],
]
`;

exports[`AUTO_LOG_LEVEL =verbose logs every stage of operation and action success of non-async function with non async sub actions 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
  Array [
    Object {
      "action": "callFunction",
      "operation": "operationFunction",
    },
  ],
  Array [
    Object {
      "action": "callFunction",
      "operation": "operationFunction",
      "result": "success",
    },
  ],
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
    },
  ],
]
`;
